src/rez/utils/request_directives.py

```diff
@@ -0,0 +1,280 @@
+from rez.vendor.version.version import VersionRange
+from rez.vendor.version.requirement import Requirement
+from rez.vendor.version.util import dewildcard
+from rez.utils.formatting import PackageRequest
+from copy import copy
+import inspect
+import sys
+
+
+# directives
+#
+
+class DirectiveBase(object):
+    """Base class of directive request handler"""
+    @classmethod
+    def name(cls):
+        """Return the name of the directive"""
+        raise NotImplementedError
+
+    def parse(self, arg_string):
+        """Parse arguments from directive syntax string"""
+        raise NotImplementedError
+
+    def to_string(self, args):
+        """Format arguments to directive syntax string"""
+        raise NotImplementedError
+
+    def process(self, range_, version, rank=None):
+        """Process requirement's version range"""
+        raise NotImplementedError
+
+
+class DirectiveHarden(DirectiveBase):
+    """Harden directive request version to specific rank"""
+    @classmethod
+    def name(cls):
+        return "harden"
+
+    def parse(self, arg_string):
+        if arg_string:
+            return [int(arg_string[1:-1].strip())]
+        return []
+
+    def to_string(self, args):
+        if args and args[0]:
+            return "%s(%d)" % (self.name(), args[0])
+        return self.name()
+
+    def process(self, range_, version, rank=None):
+        if rank:
+            version = version.trim(rank)
+        hardened = VersionRange.from_version(version)
+        new_range = range_.intersection(hardened)
+        return new_range
+
+
+# helpers
+#
+
+def parse_directive(request):
+    """Parsing directive requests when creating package
+    """
+    if "//" in request:
+        request_, directive = request.split("//", 1)
+    elif "*" in request:
+        request_, directive = _convert_wildcard_to_directive(request)
+        if not directive:
+            return request
+    else:
+        return request
+
+    # parse directive and save into anonymous inventory
+    _directive_args = directive_manager.parse(directive)
+    directive_manager.loaded.put(_directive_args,
+                                 key=request_,
+                                 anonymous=True)
+
+    return request_
+
+
+def _convert_wildcard_to_directive(request):
+    ranks = dict()
+
+    with dewildcard(request) as deer:
+        req = deer.victim
+
+        def ranking(version, rank_):
+            wild_ver = deer.restore(str(version))
+            ranks[wild_ver] = rank_
+        deer.on_version(ranking)
+
+    cleaned_request = str(req)
+    # do some cleanup
+    cleaned_request = deer.restore(cleaned_request)
+
+    if len(ranks) > 1:
+        # should we support this ?
+        return None, None
+    else:
+        rank = next(iter(ranks.values()))
+        directive = "harden(%d)" % rank
+
+        return cleaned_request, directive
+
+
+def bind_directives(package):
+    """Bind previously parsed directives to each variants
```

Comment:
Requires more explanation, what does 'bind' mean specifically?

Reply:
The `bind` is the process to retrieve previously parsed *anonymous* requirement string (see https://github.com/nerdvegas/rez/pull/1060#discussion_r625543219) and bind to the created package object as being identified. Maybe should be renamed to `identify_directives` :/